### PR TITLE
added conformance checks for port, scheme and path (extended + experimental features)

### PR DESCRIPTION
**Pull request by LiorLieberman**
**What type of PR is this?**
/kind feature
/area conformance

**What this PR does / why we need it**:
Add tests for extended redirect features (port and scheme) and path experimental feature 

**Which issue(s) this PR fixes**:

Issue #1103 

**Does this PR introduce a user-facing change?**:
NONE

### CODE REVIEW

**Comment by LiorLieberman** [comment]
@robscott

File: conformance/tests/httproute-redirect-port copy.yaml
Comment (gyohuangxin) [inline conformance/tests/httproute-redirect-port copy.yaml:1]: Is this filename `conformance/tests/httproute-redirect-port copy.yaml` correct?

File: conformance/utils/http/http.go
Comment (LiorLieberman) [inline conformance/utils/http/http.go:303]: what do you think about these lines?
I am not 100% sure it is best to take these assumptions

File: conformance/tests/httproute-redirect-port copy.yaml
Comment (LiorLieberman) [reply conformance/tests/httproute-redirect-port copy.yaml:1]: Fixed. Thanks for catching this!

File: conformance/tests/httproute-redirect-host-and-status.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-host-and-status.go:53]: The `HTTPRequestRedirectFilter` still includes the related field `Hostname`, but this is now `Host` so we have two different names for the same concept in similar areas. Can you help me to better understand the decision to change this attribute?

File: conformance/tests/httproute-redirect-port.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-port.go:31]: Port is `extended` conformance, but I don't see in this diff where we're adding a `SupportedFeature` for it? :thinking: 

See 0ddb8e68bcdbcb14df6674d20a58d270212963e8 for an example

File: conformance/tests/httproute-redirect-port.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-port.go:36]:
```suggestion
	Description: "An HTTPRoute with a port redirect filter",
```

File: conformance/tests/httproute-redirect-port.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-port.go:56]: This is a bit of a nit, but for consistency with the other tests and because this area of the tests is meant to be focused on `Port` support, should we use this name instead?

```suggestion
				Path:             "/port-and-host",
```

File: conformance/tests/httproute-redirect-port.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-port.go:78]: For completion, should we not also add a `/port-and-host-and-status` test here?

File: conformance/tests/httproute-redirect-scheme.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-scheme.go:31]: Scheme is extended, but similar to port I'm not seeing where we've added a `SupportedFeature` flag for it :thinking: 

See 0ddb8e68bcdbcb14df6674d20a58d270212963e8 for an example

File: conformance/tests/httproute-redirect-scheme.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-scheme.go:36]:
```suggestion
	Description: "An HTTPRoute with a scheme redirect filter",
```

File: conformance/tests/httproute-redirect-scheme.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-scheme.go:56]: For consistency, and since this is `Scheme` focused, should this be:

```suggestion
				Path:             "/scheme-and-host",
```

?

File: conformance/tests/httproute-redirect-scheme.go
Comment (shaneutt) [inline conformance/tests/httproute-redirect-scheme.go:78]: For completion, should we have `/scheme-and-host-and-status`?

File: conformance/utils/http/http.go
Comment (shaneutt) [inline conformance/utils/http/http.go:302]: Just for pure convention I would expect this to default to `http`, no?

File: conformance/utils/http/http.go
Comment (shaneutt) [reply conformance/utils/http/http.go:303]: The default for scheme and port made me pause slightly. Unless I'm misunderstanding things we're setting defaults here, conventionally you would (almost?) never redirect _to_ `http` so I'm just wondering if we should flip those :thinking:

**Review by shaneutt — CHANGES_REQUESTED** [review]
Looking good!

I do have a few things I would like to see addressed before we merge, but they are minor.

File: conformance/utils/http/http.go
Comment (LiorLieberman) [reply conformance/utils/http/http.go:303]: The thing is that I have the checks down there. For example the line 312 - I am checking for scheme. 
so when you would run for example a redirect test for Port when `expected.RedirectRequest.Scheme` is an empty string (because the implementations should not bother to set Scheme in their expected struct unless they want to check it) then we would fail the scheme. Hence I added it. 
My understanding is the by default it is http.

For port it is interesting, I will probably need to change it. Now I have tested Path redirect (changes are not in this PR) with envoyGateway other than istio because istio does not yet support it and I noticed that the response is an empty port, so "80" is not good.

If we decide to set the default to "80" we need to also fill `cRes.RedirectRequest.Port` to "80" but not sure thats desired.

Comment (LiorLieberman) [reply conformance/utils/http/http.go:303]: The other option I see is setting it to what we get from the expected field if it is an empty string but that also feels a bit iffy - something like this:
```go
if expected.RedirectRequest.Port == "" {
  expected.RedirectRequest.Port = cRes.RedirectRequest.Port
}
if expected.RedirectRequest.Scheme == "" {
  expected.RedirectRequest.Scheme = cRes.RedirectRequest.Scheme
}
```

I do have the same this for expected.RedirectRequest.Path which I think to add something similar:
```go
if expected.RedirectRequest.Path == "" {
  expected.RedirectRequest.Path = cRes.RedirectRequest.Path
}
```

**Comment by LiorLieberman** [comment]
@shaneutt Thanks very much for the feedback, I did miss some important things.
PTAL

File: conformance/tests/httproute-redirect-host-and-status.go
Comment (shaneutt) [reply conformance/tests/httproute-redirect-host-and-status.go:53]: I still feel a bit odd about changing this name here, but I wont consider it a blocker.

File: conformance/utils/http/http.go
Comment (shaneutt) [reply conformance/utils/http/http.go:302]: discussion of this appears to be happening in https://github.com/kubernetes-sigs/gateway-api/pull/1611#discussion_r1051991728 instead, considering this resolved.

File: conformance/utils/http/http.go
Comment (shaneutt) [reply conformance/utils/http/http.go]: @youngnick and @robscott curious as to your thoughts here.

**Review by shaneutt — APPROVED** [review]
Approving so as to no longer block as most of the issues I brought up were resolved, but would still like a review from @youngnick or @robscott for the final lgtm.

File: conformance/tests/httproute-redirect-port.go
Comment (robscott) [inline conformance/tests/httproute-redirect-port.go:51]: It looks like we default to a status code of 302 in HTTPRoute spec: 
https://github.com/kubernetes-sigs/gateway-api/blob/e097dff9f32a3edd0e9465158fa7b48028e3dc86/apis/v1beta1/httproute_types.go#L923-L925

We should expect that code to be returned here.
```suggestion
				Port: "8083",
				StatusCode: 302,
```

File: conformance/tests/httproute-redirect-port.go
Comment (robscott) [inline conformance/tests/httproute-redirect-port.go:61]: Same as above:
```suggestion
				Port: "8083",
				StatusCode: 302,
```

File: conformance/tests/httproute-redirect-port.yaml
Comment (robscott) [inline conformance/tests/httproute-redirect-port.yaml:45]: Would be useful for one of these to explicitly set 302 to ensure that works as well:
```suggestion
        statusCode: 302
```

File: conformance/tests/httproute-redirect-scheme.go
Comment (robscott) [inline conformance/tests/httproute-redirect-scheme.go:51]: Same comments as above, we should expect a 302 here and elsewhere in the test where it has not been specified. Even if we could make that a default somewhere, I'd rather explicitly state that expectation in each test case.

File: conformance/tests/httproute-redirect-scheme.yaml
Comment (robscott) [inline conformance/tests/httproute-redirect-scheme.yaml:35]: One of the status codes in this test should probably be 302 for better coverage.

**Review by robscott — COMMENTED** [review]
Thanks @LiorLieberman! A few nits but otherwise LGTM.

File: conformance/utils/http/http.go
Comment (robscott) [reply conformance/utils/http/http.go]: My preference is to avoid defaults here, specify what we care about in the test, and only check for that. In this case that would mean lots of repetitive config, but I think it would more clearly communicate what we actually expect from an implementation.

Comment (LiorLieberman) [reply conformance/utils/http/http.go]: I have made some refactoring - PTAL

**Comment by LiorLieberman** [comment]
@shaneutt @robscott 

I have made a small refactor

* Removed 302 default as per Rob's preference.
* added tests for path redirect:
** tested both with envoygateway and istio (after I have contributed the path redirect support to them(https://github.com/istio/istio/pull/42538)

* taking defaults from the original request if they arent set.

PTAL

File: conformance/utils/http/http.go
Comment (robscott) [inline conformance/utils/http/http.go:307]: Tiny nit: All the conditions above this are setting reasonable defaults, all the conditions below are checking values. I'd expect there to be more separation, maybe even comments, but in this case, this transition is the only one without an extra newline.

File: conformance/utils/suite/suite.go
Comment (robscott) [inline conformance/utils/suite/suite.go:64]:
```suggestion
	// This option indicates support for HTTPRoute path redirect (extended conformance).
```

File: conformance/tests/httproute-redirect-port.yaml
Comment (robscott) [reply conformance/tests/httproute-redirect-port.yaml:45]: I'd still like this to exist somewhere even though 302 is currently the default value, that could theoretically change in a future API version, just replacing one of the 301s with a 302 would be ideal.

**Review by robscott — COMMENTED** [review]
Thanks @LiorLieberman, a few tiny nits but otherwise LGTM.

File: conformance/utils/http/http.go
Comment (LiorLieberman) [reply conformance/utils/http/http.go:292]: added a separate function

**Comment by robscott** [comment]
Thanks @LiorLieberman!

/lgtm

**Pull request merged**
